tiledb/sm/filesystem/vfs.cc

```diff
@@ -868,19 +870,27 @@ Status VFS::read_all(
   RETURN_NOT_OK(compute_read_batches(regions, &batches));
 
   // Read all the batches and copy to the original destinations.
-  Buffer buffer;
   for (const auto& batch : batches) {
-    RETURN_NOT_OK(buffer.realloc(batch.nbytes));
-    RETURN_NOT_OK(read(uri, batch.offset, buffer.data(), batch.nbytes));
-    // Parallel copy back into the individual destinations.
-    parallel_for(0, batch.regions.size(), [&batch, &buffer](uint64_t i) {
-      const auto& region = batch.regions[i];
-      uint64_t offset = std::get<0>(region);
-      void* dest = std::get<1>(region);
-      uint64_t nbytes = std::get<2>(region);
-      std::memcpy(dest, buffer.data(offset - batch.offset), nbytes);
+    URI uri_copy = uri;
+    BatchedRead batch_copy = batch;
+    auto task = thread_pool->enqueue([uri_copy, batch_copy, this]() {
```

Comment:
I like this change, but it can deadlock if read() and read_all() use the same `ThreadPool`.

Scenario:
- `thread_pool->num_threads()` == 4
- `batches.size()` == 4
- Each of the 4 routines execute before any of them reach `read()` on line 880.
- The `thread_pool_.wait_all()` line will hang within `read()` because the tasks they are waiting on will never acquire a free thread from `thread_pool`.

There isn't a trivial fix here, other than ensuring they operate on different thread pools. So how about adding an `assert(thread_pool != &thread_pool_` here.

The universal solution to this type of deadlock is to use an async interface through the entire callstack, eliminating the use of std::future objects.

Consider the above scenario again. Instead of returning 4 std::future objects in 'tasks', we should just signal the caller exactly one time when the entire operation is complete. If the caller, executing on a thread unowned by 'thread_pool', wants to block until they receive that notification, that's fine. We can ensure within our scope (this routine, and the nested routines) that we never block by using async-only interfaces.

I understand this is a larger topic that's probably out-of-scope for this ticket. So let's either revert this or add the assert.

Reply:
Great catch on that @joe-maley -- I suggest we add the assert for this PR, and we'll file a followup issue to fix it.